lib/gcp/token_validator.go

```diff
@@ -92,5 +93,35 @@ func (id *IDTokenValidator) Validate(ctx context.Context, token string) (*IDToke
 	if err := idToken.Claims(&claims); err != nil {
 		return nil, trace.Wrap(err)
 	}
+
+	if claims.Google.ComputeEngine != (ComputeEngine{}) {
+		return &claims, nil
+	}
+
+	// If the token does not have a `google.compute_engine` claim, it is not a
+	// GCE VM token. For those tokens, we extract the project ID from the email
+	// claim.
+	// GKE Workload Identity tokens do not have the `google.compute_engine` claim
+	// and so to support Teleport services running on GKE, we need to extract the
+	// project ID from the email claim.
+	// Service accounts running on GKE are not guaranteed to have domain emails
+	// with the following format: <sa_name>@<project_id>.iam.gserviceaccount.com
+	// See https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity#authenticating_to_other_services
+	// See https://cloud.google.com/iam/docs/service-accounts#service_account_email_addresses
+	// See https://cloud.google.com/iam/docs/service-account-types#user-managed
+	matches := gcpServiceAccountEmailRegex.FindStringSubmatch(claims.Email)
+	if len(matches) != 2 {
+		return nil, trace.BadParameter("invalid email claim: %q", claims.Email)
+	}
+
+	// Assign the project ID exatracted from the email to the Google claims.
+	claims.Google.ComputeEngine.ProjectID = matches[1]
+
 	return &claims, nil
 }
+
+// gcpServiceAccountEmailRegex is the regex used to extract the project ID from
+// the email claim of a GCP service account.
+// See https://cloud.google.com/iam/docs/service-accounts#service_account_email_addresses
+// See https://cloud.google.com/iam/docs/service-account-types#user-managed
+var gcpServiceAccountEmailRegex = regexp.MustCompile(`(?s)^[^@]+@([^\.]+)\.iam\.gserviceaccount\.com$`)
```

Comment:
Can we name this to make it clear this is only for user managed service accounts? 

This won't match for the default compute service account which looks like `project-number-compute@developer.gserviceaccount.com`.

This also raises the question, should we support the use of the default compute service account ? It's definitely poor security practice for customers to be using this - but I wouldn't be surprised if someone tries it - perhaps we can log an error/warning that at least explains that we don't support the default compute service account ?

Reply:
I deliberately removed the support for the default compute service account.

Added an error rejecting the default service account usage

Reply:
Perfect - we want to guide people away from that anyway.

Reply:
@tigrato If we're removing support for the default service account, we'll need to update the docs to reflect that.

Reply:
@atburke I only removed the support if it's not running in a VM  but I can document it 

@strideynet should we also consider removing support from VM instances? I am afraid that someone else already uses it